### ISSUE TITLE
Performance issue

### ISSUE THREAD

**Issue by awav** [google/jax#2419]
Hello everyone,

I'm testing JAX framework speed and for that purpose, I wrote the **Kalman filter**. The gold standard is a Matlab version that performs an algorithm for 50k number of steps in ~0.1 sec. The TensorFlow 2.1 with `tf.function` gives ~2.0 sec. **Unfortunately, JAX version hangs forever**.

The JAX version of Kalman filter in Colab is [here](https://colab.research.google.com/drive/16Ko_vHrE6Wg7J_hgmglpv_wQYx5q7vlv) and TensorFlow 2.1 version is [here](https://colab.research.google.com/drive/133FyNrUNlqj8GFgsf8G7T8L9TY9wKcay). I would appreciate any help in resolving the issue or understanding what is wrong with the code.

Thanks,
Artem

**Comment by martinsotir** [comment]
Hi @awav,

It seems,. Jax JIT unrolls for loops and the compilation time increases with the number of iterations up to the point where it never finishes compiling (with N=50k).

Instead, you could use `jax.lax.scan` to perform the loop (`scan()` is similar to a reduce in the functional paradigm).

See also [the control flow section in the Jax documentation](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Control-Flow).

Here what I have tried ([full notebook here](https://colab.research.google.com/drive/16tyHTH5Rl1FmDr4YGaIzN0KMaUk-wCk5)):

```python
def kalman_filter_scan(y_observed, var_likelihood, var_prior, len_prior, dtype=None):
    dtype = jnp.float32 if dtype is None else dtype
    lam = jnp.sqrt(5.0) / len_prior
    F = jnp.array(
        [[0.0, 1.0, 0.0], [0.0, 0.0, 1.0], [-(lam ** 3), -3 * lam ** 2, -3 * lam]], dtype=dtype
    )
    H = jnp.array([[1.0, 0.0, 0.0]], dtype=dtype)
    kappa = 5.0 / 3.0 * var_prior / len_prior ** 2
    Pinf = jnp.array(
        [
            [var_prior, 0.0, -kappa],
            [0.0, kappa, 0.0],
            [-kappa, 0.0, 25.0 * var_prior / len_prior ** 4.0],
        ],
        dtype=dtype,
    )
    m = jnp.zeros([F.shape[0], 1], dtype=dtype)
    P = Pinf
    A = jscipy.linalg.expm(F)
    log_marg_lik = 0.0

    def inner_loop(carry, y_observed_k):
      m, P, log_marg_lik = carry

      m = A @ m
      P = A @ (P - Pinf) @ A.T + Pinf
      S = var_likelihood + H @ P @ H.T
      v = y_observed_k - H @ m
      v2 = jnp.square(v)
      lZ = -0.5 * (v2 / S[0, 0] + jnp.log(2 * jnp.pi * S))
      log_marg_lik += lZ[0, 0]
      K = P @ H.T / S[0, 0]
      m = m + K @ v
      P = P - K @ S @ K.T

      return (m, P, log_marg_lik), None

    (_, _, log_marg_lik), _ = jax.lax.scan(inner_loop, (m, P, log_marg_lik), y_observed)


    return log_marg_lik
```

I checked that the returned scalars where the same in both versions, but I must certainly have missed something as I'm getting very optimistic results with 128us per loop for the jitted version and 202 ms / loop un-jitted.

**Comment by awav** [comment]
@martinsotir, thanks a lot! Indeed, the timing is promising. Although, I don't know what CPU/RAM/... used by Colab and it is not really a fair comparison with Matlab.

> Jax JIT unrolls for loops and the compilation time increases with the number of iterations up to the point where it never finishes compiling (with N=50k).
> 
> Instead, you could use jax.lax.scan to perform the loop (scan() is similar to a reduce in the functional paradigm).

With a message like "JAX is NumPy on steroids" that can be a problem. Guess, my question is why would JAX try to unroll `for` loop, isn't that an issue?

**Comment by skye** [comment]
I think you need to use `block_until_ready` in your colab microbenchmark. Something like:
```python
%timeit -n10 kalman_filter_jit(y, var_y, var_f, len_f).block_until_ready()
```
This is because JAX dispatches operations asynchronously (usually a good thing!), so you're only measuring the dispatch time here. See https://jax.readthedocs.io/en/latest/async_dispatch.html for more details.

Regarding the unrolled for-loop, this is an artifact of how JAX traces functions to be jitted. It's a little nuanced, but basically, JAX only "sees" certain overloaded operations (e.g. jax.numpy functions), and Python doesn't provide a good way to overload control flow, such as for-loops. The result is the loop runs as usual (JAX doesn't "see" it at all) and JAX traces the operations inside each iteration, but can't tell they're all from the same loop body. That's why we provide control flow operations like `scan`, so JAX can "see" the loop itself. [This talk](https://slideslive.com/38923687/jax-accelerated-machinelearning-research-via-composable-function-transformations-in-python) goes over JAX's tracing in more detail if you're curious.

**Comment by awav** [comment]
> I think you need to use `block_until_ready` in your colab microbenchmark. Something like:
> 
> ```python
> %timeit -n10 kalman_filter_jit(y, var_y, var_f, len_f).block_until_ready()
> ```
> 
> This is because JAX dispatches operations asynchronously (usually a good thing!), so you're only measuring the dispatch time here. See https://jax.readthedocs.io/en/latest/async_dispatch.html for more details.



> I think you need to use `block_until_ready` in your colab microbenchmark. Something like:
> 
> ```python
> %timeit -n10 kalman_filter_jit(y, var_y, var_f, len_f).block_until_ready()
> ```
> 
> This is because JAX dispatches operations asynchronously (usually a good thing!), so you're only measuring the dispatch time here. See https://jax.readthedocs.io/en/latest/async_dispatch.html for more details.

Thanks! Good to know about that. Originally, I discovered the slow behaviour on my laptop and PC and then I ported it to the colab notebook, so I believe it is not a colab's side issue.

**Comment by awav** [comment]
> This talk goes over JAX's tracing in more detail if you're curious.

The link to slides doesn't work.

**Comment by skye** [comment]
Huh, it works for me... https://slideslive.com/38923687/jax-accelerated-machinelearning-research-via-composable-function-transformations-in-python, it's a link to a video actually. Here are just the slides: https://program-transformations.github.io/slides/NeurIPS_workshop_JAX_talk.pdf, although I'm not sure if they make sense without the commentary.

**Comment by martinsotir** [comment]
> I think you need to use block_until_ready in your colab microbenchmark. 

Thanks! I fixed the notebook. And am getting a more realistic 62ms/loop.

I initially though that the `block_until_ready` was not required on CPU but I learned that Jax always dispatch asynchronously ! This behaviour is [well documented](https://jax.readthedocs.io/en/latest/async_dispatch.html) but it could be a small gotcha for pytorch users that are used to async. execution on GPU vs synch. execution on CPU.
